### DISCUSSION TITLE
Why doesn't Valtio automatically track dependencies?

### DISCUSSION THREAD

**Discussion by transitive-bullshit** [pmndrs/valtio#950] [General]
During my [deep-dive comparison of reactive TS frameworks](https://github.com/transitive-bullshit/ts-reactive-comparison), I was surprised to see that Valtio doesn't automatically track dependency reads in its computed properties or `subscribe` (which seems like the closest analog Valtio has to an `effect`).

https://github.com/pmndrs/valtio/issues/149 seems to have settled on a compromise by adding an optional `watch` to `valtio/utils` which at least makes dependency tracking possible, albeit with a somewhat awkward manual DX compared to other similar reactive libs ([example](https://github.com/transitive-bullshit/js-reactivity-benchmark/blob/feature/minimal-issues-repro-valtio/src/frameworks/valtio.ts) which I may or may not have butchered 😂 ).

I know the authors of this project likely have a very good reason for this design decision, but I wasn't able to understand it by searching through the repo and old issues.

Related: https://github.com/pmndrs/valtio/issues/149 and potentially #82 

I'd love to update my comparison with an explanation of how you guys think about automatic dependency tracking versus the more explicit approach taken by libs like Valtio and Jotai.

Thanks!

**Comment by dai-shi** [comment]
Because it's not the goal nor the scope. `watch` hasn't be actively improved lately and it may not provide dependency tracking mechanism. I think we need to overhaul `watch` to match your requirement (which sounds interesting).

fyi, the goal of valtio is to create immutable snapshots.
```js
const p = proxy({ count: 1 });
const s1 = snapshot(p);
p.count = 2;
const s2 = snapshot(p);
// s1 should be { count: 1 } and s2 should be { count: 2 }
```

**Reply by transitive-bullshit** [reply]
Makes sense; thanks for the super quick reply, Daishi.

Would it be more accurate then to compare Valtio to Immer, as opposed to signals frameworks?

**Reply by dai-shi** [reply]
Yeah, that's what I was gonna say: Valtio is more comparable with Immer and ImmutableJS.

One of my old tweets: https://x.com/dai_shi/status/1330490462642192384

![image](https://github.com/user-attachments/assets/52b7530a-7107-48f9-8d8d-a9cfd0d70b27)

**Reply by transitive-bullshit** [reply]
Makes sense 🙏
